internal/pkg/util/fs/overlay/overlay_item_linux.go

```diff
@@ -0,0 +1,297 @@
+// Copyright (c) 2019-2023, Sylabs Inc. All rights reserved.
+// This software is licensed under a 3-clause BSD license. Please consult the
+// LICENSE.md file distributed with the sources of this project regarding your
+// rights to use or distribute this software.
+
+package overlay
+
+import (
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"syscall"
+
+	"github.com/sylabs/singularity/internal/pkg/util/bin"
+	"github.com/sylabs/singularity/pkg/image"
+	"github.com/sylabs/singularity/pkg/sylog"
+)
+
+// Item represents information about a single overlay item (as specified,
+// for example, in a single --overlay argument)
+type Item struct {
+	// Type represents what type of overlay this is (from among the values in pkg/image)
+	Type int
+
+	// Writable represents whether this is a writable overlay
+	Writable bool
+
+	// BarePath is the path of the overlay stripped of any colon-prefixed
+	// options (like ":ro")
+	BarePath string
+
+	// DirToMount is the path of the directory that will actually be passed to
+	// the mount system-call when mounting this overlay
+	DirToMount string
+
+	// secureParentDir is the (optional) location of a secure parent-directory
+	// in which to create mount directories if needed. If empty, one will be
+	// created with os.MkdirTemp()
+	secureParentDir string
+}
+
+// NewOverlayFromString takes a string argument, as passed to --overlay, and returns
+// an overlayInfo struct describing the requested overlay.
+func NewOverlayFromString(overlayString string) (*Item, error) {
+	overlay := Item{Writable: true}
+
+	splitted := strings.SplitN(overlayString, ":", 2)
+	overlay.BarePath = splitted[0]
+	if len(splitted) > 1 {
+		if splitted[1] == "ro" {
+			overlay.Writable = false
+		}
+	}
+
+	s, err := os.Stat(overlay.BarePath)
+	if (err != nil) && os.IsNotExist(err) {
+		return nil, fmt.Errorf("specified overlay %q does not exist", overlay.BarePath)
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	if s.IsDir() {
+		overlay.Type = image.SANDBOX
+	} else if err := analyzeImageFile(&overlay); err != nil {
+		return nil, fmt.Errorf("error encountered while examining image file %s: %s", overlay.BarePath, err)
+	}
+
+	return &overlay, nil
+}
+
+// analyzeImageFile attempts to determine the format of an image file based on
+// its header
+func analyzeImageFile(overlay *Item) error {
```

Comment:
This should be modified to be `func (i *Item) analyzeImageFile() error`. Using a method on the struct is nicer than a separate function that modifies a struct passed by pointer, given that the code is entirely concerned with an `Item`.